### PR TITLE
Clarify docs on accumulating into arrays in prange

### PR DESCRIPTION
**Pull request by synapticarbors**
Seeing a [question on stackoverflow](https://stackoverflow.com/questions/56556416), reminded me of an issue a posed a while back (#3681). I've updated the docs to hopefully make this clearer since it seems to be potentially common point of confusion. If my additional notes on the issue or example could be further improved, please let me know.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. I think this documentation is very helpful. I did have a quick hack at the source and I think it's probably going to be possible in the long term to detect such cases. An incredibly specific to the example case patch: https://github.com/stuartarchibald/numba/commit/50d6335e3f02ee12d6da21954d434658149ab9b1
Gives:
```
<path>/numba/numba/npyufunc/parfor.py:762: NumbaParallelSafetyWarning: Variable $30.3 used in parallel loop may be written to simultaneously by multiple workers and may result in non-deterministic or unintended results.

File "issue4170.py", line 7:
def prange_wrong_result(x):
    <source elided>
    y = np.zeros(4)
    for i in prange(n):
    ^

  warnings.warn(NumbaParallelSafetyWarning(msg, loc))
Race condition on: i % $const32.4
[1971. 1958. 1964. 1961.]
```

**Comment by stuartarchibald** [comment]
@DrTodd13 any thoughts on this? Thanks.

File: docs/source/user/parallel.rst [hunk 1]
```diff
@@ -84,7 +84,11 @@ function/operator using its previous value in the loop body. The initial value
 of the reduction is inferred automatically for ``+=`` and ``*=`` operators.
 For other functions/operators, the reduction variable should hold the identity
 value right before entering the ``prange`` loop.  Reductions in this manner
-are supported for scalars and for arrays of arbitrary dimensions.
+are supported for scalars and for arrays of arbitrary dimensions. Care should be
```
Comment (DrTodd13) [inline docs/source/user/parallel.rst]: I think it would be better if this section on what not to do was moved to right before the negative example you included below.

File: docs/source/user/parallel.rst [hunk 2]
```diff
@@ -84,7 +84,11 @@ function/operator using its previous value in the loop body. The initial value
...
+taken, however, as reducing into slices or elements of an array is not supported 
+if the elements specified by the slice or index is written to simultaneously by 
+multiple parallel threads. The compiler will not detect such cases and may result
```
Comment (DrTodd13) [inline docs/source/user/parallel.rst]: The compiler does have some code for race detection so I might reword this as "The compiler may not detect such cases  and then a race condition would occur."

**Comment by DrTodd13** [comment]
Also, this example is very specific.  If possible, maybe some more text listing a few more examples of the kind of mistakes people make and then you can show your example but making sure to indicate it is just an example mistake and not the only one possible.

**Comment by synapticarbors** [comment]
@DrTodd13 -- thank you for the suggests. I updated the PR in response. I wasn't sure what other types of mistakes are common when using `prange`, but the example captures one that I made and saw repeated on stackoverflow. I'm happy to add other examples, but I wasn't 100% sure what you were getting at.

**Comment by stuartarchibald** [comment]
This sort of thing is a demonstration of slicing not working:
```python
from numba import njit, prange
import numpy as np

@njit(parallel=True)
def prange_wrong_result(x):
    n = x.shape[0]
    y = np.zeros((4, 1))
    for i in prange(n):
        y[:] += x[i]

    return y


z = np.arange(10000)
print(prange_wrong_result(z))
```

whereas the whole array reduction is fine:

```python
from numba import njit, prange
import numpy as np

@njit(parallel=True)
def prange_wrong_result(x):
    n = x.shape[0]
    y = np.zeros((4, 1))
    for i in prange(n):
        y += x[i]
    return y


z = np.arange(10000)
print(prange_wrong_result(z))
```
as is creating a slice ref outside the parallel reduction loop and using that:
```python
from numba import njit, prange
import numpy as np

@njit(parallel=True)
def prange_wrong_result(x):
    n = x.shape[0]
    y = np.zeros((4, 1))
    z = y[:]
    for i in prange(n):
        z += x[i]
    return y

z = np.arange(10000)
print(prange_wrong_result(z))
```

**Comment by synapticarbors** [comment]
@stuartarchibald - So just to be clear, you would like me to put those 3 examples in the PR at the end of the section?

**Comment by stuartarchibald** [comment]
@DrTodd13 any view on the above?  @synapticarbors The first one, which is a common issue, probably should go in, I'm kind of edging towards including the other two as well as a demonstration of a subtle change that can make it work fine.

**Comment by DrTodd13** [comment]
@stuartarchibald @synapticarbors Yes, I'd say add them all.

**Comment by synapticarbors** [comment]
I updated the section with all of the additional examples. I left in a version with the modulo operator since I thought that felt like a different use case than whole array or slice reduction and might be more illustrative for the user.

File: docs/source/user/parallel.rst [hunk 3]
```diff
@@ -84,7 +84,7 @@ function/operator using its previous value in the loop body. The initial value
 of the reduction is inferred automatically for ``+=`` and ``*=`` operators.
 For other functions/operators, the reduction variable should hold the identity
 value right before entering the ``prange`` loop.  Reductions in this manner
-are supported for scalars and for arrays of arbitrary dimensions.
+are supported for scalars and for arrays of arbitrary dimensions.
```
Comment (stuartarchibald) [inline docs/source/user/parallel.rst]:
```suggestion
are supported for scalars and for arrays of arbitrary dimensions.
```

File: docs/source/user/parallel.rst [hunk 4]
```diff
@@ -116,6 +116,71 @@ The following example demonstrates a product reduction on a two-dimensional arra
...
+        n = x.shape[0]
+        y = np.zeros(4)
+        for i in prange(n):
+            # accumulating into the same element of `y` from different
+            # parallel iterations of the loop results in a race condition
+            y[:] += x[i]
+
+        return y
+
+as does the following example where explicitly specifying the accumulating element::
```
Comment (stuartarchibald) [inline docs/source/user/parallel.rst]:
```suggestion
as does the following example where the accumulating element is explicitly specified::
```
perhaps ^ ?

**Comment by seibert** [comment]
Merging now that the docs test has passed.
